File: 05-Algorithms/code/typescript/src/bipartite.ts
import { buildAdjList, type Edge } from "./graphCommon.js";

/**
 * Checks whether the (undirected) graph on n vertices is 2-colorable: BFS
 * from every uncolored vertex, alternating colors across edges, and
 * failing as soon as an edge connects two same-colored vertices (including
 * a self-loop, which is always odd-length and therefore never bipartite).
 * O(V + E).
 */
export function isBipartite(n: number, edges: Edge[]): boolean {
  const adj = buildAdjList(n, edges, false);
  const color = new Array(n).fill(-1);
  for (let start = 0; start < n; start++) {
    if (color[start] !== -1) continue;
    color[start] = 0;
    const queue = [start];
    let head = 0;
    while (head < queue.length) {
      const u = queue[head++];
      for (const v of adj[u]) {
        if (color[v] === -1) {
          color[v] = 1 - color[u];
          queue.push(v);
        } else if (color[v] === color[u]) {
          return false;
        }
      }
    }
  }
  return true;
}

File: 05-Algorithms/code/python/seb_alg/bipartite.py
"""Mirrors `typescript/src/bipartite.ts` exactly."""

from __future__ import annotations

from typing import Sequence

from .graph_common import build_adj_list


def is_bipartite(n: int, edges: Sequence[Sequence[int]]) -> bool:
    """Checks whether the undirected graph is 2-colorable via BFS. O(V + E)."""
    adj = build_adj_list(n, edges, False)
    color = [-1] * n
    for start in range(n):
        if color[start] != -1:
            continue
        color[start] = 0
        queue = [start]
        head = 0
        while head < len(queue):
            u = queue[head]
            head += 1
            for v in adj[u]:
                if color[v] == -1:
                    color[v] = 1 - color[u]
                    queue.append(v)
                elif color[v] == color[u]:
                    return False
    return True
